shiny future: Barbara makes a wish (#138)
rust-lang/wg-async-foundations: src/vision/shiny_future/barbara_makes_a_wish.md

## The story

[Barbara] has an initial prototype of a new service she wrote in sync Rust. She then decides, since the service is extremely I/O bound, to port it to async Rust and her benchmarks have led her to believe that performance is being left on the table.

She does this by sprinkling `async/.await` everywhere, picking an executor, and moving dependencies from sync to async.

Once she has the program compiling, she thinks "oh that was easy". She runs it for the first time and surprisingly she finds out that when hitting an endpoint, nothing happens.

Barbara, always prepared, has already added logging to her service and she checks the logs. As she expected, she sees here that the endpoint handler has been invoked but then... nothing. Barbara exclaims, "Oh no! This was not what I was expecting, but let's dig deeper." 

She checks the code and sees that the endpoint spawns several tasks, but unfortunately those tasks don't have much logging in them.

> Barbara now remembers hearing something about an `wish4-async-insight` crate, which has gotten some buzz on her Rust-related social media channels. She decides to give that a shot.

Reviewer: ```suggestion
Barbara now remembers hearing something about a `wish4-async-insight` crate, which has gotten some buzz on her Rust-related social media channels. She decides to give that a shot.
```

She adds the crate as a dependency to her `Cargo.toml`, renaming it to just `insight` to make it easier to reference in her code, and then initializes it in her main async function.

```rust,ignore
async fn accept_loop(addr: impl ToSocketAddrs) -> Result<()> {
    insight::init(); // new code
    ...
}
```

> Barbara rebuilds and run her program again. She doesn't see anything different in the terminal output for the program itself though, and the behavior is the same as before: hitting an endpoint, nothing happens. She double-checks the readme for the `async-executor-insight` crate, and realizes that she needs to connect other programs to her service to observe the insights being gathered. Barbara decides that she wants to customize the port that `insight` is listening on before she starts her experiments with those programs.

Reviewer: ```suggestion
Barbara rebuilds and runs her program again. She doesn't see anything different in the terminal output for the program itself though, and the behavior is the same as before: hitting an endpoint, nothing happens. She double-checks the readme for the `async-executor-insight` crate, and realizes that she needs to connect other programs to her service to observe the insights being gathered. Barbara decides that she wants to customize the port that `insight` is listening on before she starts her experiments with those programs.
```

> Barbara rebuilds and run her program again. She doesn't see anything different in the terminal output for the program itself though, and the behavior is the same as before: hitting an endpoint, nothing happens. She double-checks the readme for the `async-executor-insight` crate, and realizes that she needs to connect other programs to her service to observe the insights being gathered. Barbara decides that she wants to customize the port that `insight` is listening on before she starts her experiments with those programs.

Author: :rofl: because I renamed the crate while I was in the middle of writing this, and didn't do a proper search-replace.

Will fix, probably on monday!

> Barbara rebuilds and run her program again. She doesn't see anything different in the terminal output for the program itself though, and the behavior is the same as before: hitting an endpoint, nothing happens. She double-checks the readme for the `async-executor-insight` crate, and realizes that she needs to connect other programs to her service to observe the insights being gathered. Barbara decides that she wants to customize the port that `insight` is listening on before she starts her experiments with those programs.

Author: fixed in 0a023c2e8c73d5f6b06f41f6faf257b1b06ebca0

```rust,ignore
async fn accept_loop(addr: impl ToSocketAddrs) -> Result<()> {
    insight::init(listen_port => 8080); // new code, leveraging keyword arguments feature added in 2024
    ...
}
```

While her code rebuilds, Barbara investigates what programs she might use to connect to the insight crate.

One such program, `consolation`, can run in the terminal. Barbara is currently just deploying her service locally on her development box, so she opts to try that out and see what it tells her.

```
% rustup install wish4-consolation
...
% consolation --port 8080
```

This brings up a terminal window that looks similar to the Unix `top` program, except that instead of a list of OS processes, this offers a list of tasks, with each task having a type, ID, and status history (i.e. percentage of time spent in running, ready to poll, or blocked). Barbara skims the output in the list, and sees that one task is listed as currently blocked.

> Barbara taps the arrow-keys and sees that this causes a cursor to highlight different tasks in the list. She highlights the blocked task and hits the Enter key. This cause the terminal to switch to a Task view, describing more details about that task and its status.

Reviewer: ```suggestion
Barbara taps the arrow-keys and sees that this causes a cursor to highlight different tasks in the list. She highlights the blocked task and hits the Enter key. This causes the terminal to switch to a Task view, describing more details about that task and its status.
```

The Task view here says that the task is blocked, references a file and line number, and also includes the line from the source code, which says `chan.send(value).await`. The blocked task also lists the resources that the task is waiting on: `prototype_channel`, and next to that there is text on a dark red background: "waiting on channel capacity." Again, Barbara taps the arrow-keys and sees that she can select the line for the resource.

> Barbara notices that this whole time, at the bottom of the terminal, there was a line that says "For help, hit `?` key"; she taps question mark. This brings up a help message in a scrollable subwindow explaining the task view in general, sa well as link to online documentation. The help message notes that the user can follow the chain: One can go from the blocked task to the resource its waiting on, and from that resource to a list of tasks responsible for freeing up the resource.

Reviewer: ```suggestion
Barbara notices that this whole time, at the bottom of the terminal, there was a line that says "For help, hit `?` key"; she taps question mark. This brings up a help message in a scrollable subwindow explaining the task view in general as well as link to online documentation. The help message notes that the user can follow the chain: One can go from the blocked task to the resource it's waiting on, and from that resource to a list of tasks responsible for freeing up the resource.
```

> Barbara notices that this whole time, at the bottom of the terminal, there was a line that says "For help, hit `?` key"; she taps question mark. This brings up a help message in a scrollable subwindow explaining the task view in general, sa well as link to online documentation. The help message notes that the user can follow the chain: One can go from the blocked task to the resource its waiting on, and from that resource to a list of tasks responsible for freeing up the resource.

Author: addressed in a64f8ac2aebb91af36ad9c634727352ffad636e1, thanks @Stupremee !

...
% consolation --port 8080
```

This brings up a terminal window that looks similar to the Unix `top` program, except that instead of a list of OS processes, this offers a list of tasks, and shows their status (i.e. running, ready to poll, or blocked), as well as some metrics about how long the tasks spend in each state.

At the top of the screen, Barbara sees highlighted warning: "deadlock cycle was detected. hit `P` for more info."

Barbara types capital `P`. The terminal switches to "problem view," which shows

 * The task types, ID, and attributes for each type.
 * The resources being awaited on
 * The location / backtrace of the await.
 * A link to a documentation page expanding on the issue.

The screen also says "hit `D` to generate a graphviz `.dot` file to disk describing the cycle."

Barbara hits `D` and stares at the resulting graph.

Barbara suddenly realizes her mistake: She had constructed a single task that was sometimes enqueuing work (by sending messages on the channel), and sometimes dequeuing work, but she had not put any controls into place to ensure that the dequeuing (via `recv`) would get prioritized as the channel filled up.

> Barbara reflects on the matter: she knows that she could swap in an unbounded channel to resolve this, but she think that she would be better off thinking a bit more about her system design, to see if she can figure out a way to supply back-pressure so that the send rate will go down as the channel fills up.

Reviewer: ```suggestion
Barbara reflects on the matter: she knows that she could swap in an unbounded channel to resolve this, but she thinks that she would be better off thinking a bit more about her system design, to see if she can figure out a way to supply back-pressure so that the send rate will go down as the channel fills up.
```

## 🤔 Frequently Asked Questions

### **What status quo story or stories are you retelling?**

[Barbara wants Async Insights](../status_quo/barbara_wants_async_insights.md)

### **What is [Alan] most excited about in this future? Is he disappointed by anything?**

Alan is happy to see a tool that gives one a view into the internals of the async executor.

> Alan is not so thrilled about using the `consolation` terminal interface; but luckily there are other options, namely a web-browser based client that offers even richer functionality, such as renderings of the task/resource dependency graph.

Reviewer 2: Can we also mention integrated editor extensions that provide this functionality as well?

Author: Fixed in 0eaf2add9434d720ce0732df07e6ded8188e3f2c

### **What is [Grace] most excited about in this future? Is she disappointed by anything?**

Grace is happy to see a tool, but wonders whether it could have been integrated into `gdb`.

Grace is not so thrilled to learn that this tool is not going to try to provide specific insight into performance issues that arise solely from computational overheads in her own code. (The readme for `wish4-async-insight` says on this matter "for that, use perf," which Grace finds unsatisfying.)

### **What is [Niklaus] most excited about in this future? Is he disappointed by anything?**

> Niklaus is happy to learn that the `wish4-async-insight` is supported by both `async-std` and `tokio`, since he relies on friends in both communities to help him learn more about Async Rust.

Reviewer 2: This might not be possible to answer, but do we have a feeling for how likely it would be for alternative runtimes (e.g., specialized, bespoke runtimes) to integrate with this tool?

Author: Fixed in 671e42dfefd7832ece984b8b8e71e7fc7037cb38

Niklaus is happy about the tool's core presentation oriented around abstractions he understands (tasks and resources). Niklaus is also happy about the integrated help.

However, Niklaus is a little nervous about some of the details in the output that he doesn't understand.

### **What is [Barbara] most excited about in this future? Is she disappointed by anything?**

> Barbara is thrilled with

Reviewer 2: Looks like this isn't finished here 😊

Author: Fixed in 671e42dfefd7832ece984b8b8e71e7fc7037cb38
